### PR TITLE
feat(BOUN-1168): add dynamic route provider (Discovery Library)

### PR DESCRIPTION
**Pull request by nikolay-komarevskiy**
# Description

API boundary nodes (BNs) [have already been](https://dashboard.internetcomputer.org/nodes?type=ApiBoundary) introduced into the Internet Computer (IC) landscape. Currently, the `agent` does not support dynamic discovery of existing API BNs or the dynamic routing of requests to these nodes based on specific strategies, such as latency. This PR enhances the `agent` by adding functionality for the dynamic routing to API BNs. In a nutshell the newly introduced [DynamicRouteProvider](https://github.com/dfinity/agent-rs/pull/568/files#diff-c3cffa98e93a87951c04778073900f72c8e3b8225b33b30a9ed2cc88545bb3c8R51)  is the implementation of the [RouteProvider trait](https://github.com/dfinity/agent-rs/blob/be929fd7967249c879f48f2f494cbfc5805a7d98/ic-agent/src/agent/http_transport/route_provider.rs#L17).

Example usage 1 - ergonomic:
```sh
#[tokio::main]
async fn main() {
    let client = Client::builder().build().expect("failed to build a client");

    // Initialize the agent with a custom dynamic transport layer
    // NOTE: this requires an async context
    let agent = Agent::builder()
        .with_discovery_transport(client)
        .await
        .build()
        .expect("failed to create an agent");
    //
    // ... use agent
    //

}
```

Example usage 2 - fully customized:
```sh
#[tokio::main]
async fn main() {
    let client = Client::builder().build().expect("failed to build a client");
    let seeds = vec![Node::new("ic0.app").expect("domain is invalid")];
    let snapshot = LatencyRoutingSnapshot::new();
    let route_provider = DynamicRouteProviderBuilder::new(snapshot, seeds, client.clone()).build().await;
    let transport = ReqwestTransport::create_with_client_route(Arc::new(route_provider), client.clone()).unwrap();

    // Initialize the agent with a custom dynamic transport layer
    let agent = Agent::builder()
        .with_transport(transport)
        .build()
        .expect("failed to create an agent");
    //
    // ... use agent
    //

}
```

# How Has This Been Tested?

1. Added a [mainnet test](https://github.com/dfinity/agent-rs/pull/568/files#diff-c3cffa98e93a87951c04778073900f72c8e3b8225b33b30a9ed2cc88545bb3c8R344)
2. Created a [draft PR](https://github.com/dfinity/ic/pull/516) and [tested](https://dash.zh1-idx1.dfinity.network/invocation/3d9862ab-04c2-431e-b189-3dd506a2bc4f?target=%2F%2Frs%2Ftests%2Fboundary_nodes%3Aapi_boundary_node_decentralization_test&targetStatus=5) on the [api_boundary_node_decentralization_test](https://sourcegraph.com/github.com/dfinity/ic/-/blob/rs/tests/boundary_nodes/BUILD.bazel?L155), which is a `system-test` in the `ic-repo`.
**NOTE**: @adamspofford-dfinity and @lwshang the suggestion to fully [remove](https://github.com/dfinity/agent-rs/pull/568/commits/ea4c8e5d9e908d45147689e90df2604d30742c61) the `fetch_root_key()` was probably wrong. As `system-test` fails without this call. Could You commit a conditional `fetch_root_key()` invocation, which would work for both `mainnet` and `testnets`.

# Checklist:

- [x] The title of this PR complies with [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/).
- [ ] I have edited the CHANGELOG accordingly.
- [x] I have made corresponding changes to the documentation.

### CODE REVIEW

File: ic-agent/src/agent/http_transport/dynamic_routing/node.rs [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+}
+
+impl From<&ApiBoundaryNode> for Node {
+    fn from(api_bn: &ApiBoundaryNode) -> Self {
+        Node::new(api_bn.domain.as_str()).unwrap()
+    }
+}
+
+pub fn is_valid_domain<S: AsRef<str>>(domain: S) -> bool {
+    // TODO
```
Comment (r-birkner) [inline ic-agent/src/agent/http_transport/dynamic_routing/node.rs:41]: I would just use the same as here: https://github.com/dfinity/ic/blob/master/rs/registry/canister/src/mutations/common.rs#L131-L155

File: ic-agent/src/agent/http_transport/dynamic_routing/dynamic_route_provider.rs [hunk 2]
```diff
@@ -0,0 +1,596 @@
...
+        })?;
+        Ok(node.to_routing_url())
+    }
+}
+
+impl<S> DynamicRouteProvider<S>
+where
+    S: RoutingSnapshot + 'static,
+{
+    pub fn new(snapshot: S, seeds: Vec<Node>, http_client: Client) -> Self {
```
Comment (r-birkner) [inline ic-agent/src/agent/http_transport/dynamic_routing/dynamic_route_provider.rs:78]: This means that to create an agent with the discovery library, you need to do something like:
```
let snapshot = LatencyBasedSnapshot::new();
let http_client = Client::builder().build().unwrap();
let route_provider = HealthCheckRouteProvider::new(
            snapshot,
            vec![Node::new(domain_seed)],
            http_client
);
let transport = ReqwestTransport::create_with_client_route(
        Arc::clone(&route_provider) as Arc<dyn RouteProvider>,
        http_client,
    )
    .expect("failed to create transport");
    
let agent = Agent::builder()
        .with_transport(transport)
        .build()?;
```

Can't we do something like and just provide defaults:
```
let agent = Agent::builder()
        .with_latency_based_routing()
        .build()?;
```
or
```
let agent = Agent::builder()
        .with_health_based_routing()
        .build()?;
```

Comment (r-birkner) [reply ic-agent/src/agent/http_transport/dynamic_routing/dynamic_route_provider.rs:78]: If someone wants not the default, they can then go through the process above and create their own snapshot/route provider, etc.

File: ic-agent/src/agent/http_transport/dynamic_routing/node.rs (hunk 1 above)
Comment (nikolay-komarevskiy) [reply ic-agent/src/agent/http_transport/dynamic_routing/node.rs]: i think reusing the `Url` library some more concise:
```
pub fn is_valid_domain<S: AsRef<str>>(domain: S) -> bool {
    // Prepend scheme to make it a valid URL
    let url_string = format!("http://{}", domain.as_ref());
    Url::parse(&url_string).is_ok()
}
```

File: ic-agent/src/agent/http_transport/dynamic_routing/dynamic_route_provider.rs (hunk 2 above)
Comment (nikolay-komarevskiy) [reply ic-agent/src/agent/http_transport/dynamic_routing/dynamic_route_provider.rs:78]: Eventually this is the direction to go, yes! But i don't want to change the `agent` during the library migration phase., better in the follow up.

**Review by adamspofford-dfinity — APPROVED** [review]

**Pull request merged**